Add a CockroachDB PetSet example

tbg opened the pull request:
The example starts a simple five-node cluster with otherwise
default settings (in particular, 3x replication).

cc @bprashanth

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/28446)

<!-- Reviewable:end -->

bprashanth commented on examples/cockroachdb/README.md:
````diff
@@ -0,0 +1,19 @@
+
+When all (or enough) nodes are up, simulate a failure like this:
+
+```shell
````
please provide an example of how to set/get a key:value. An example displaying the replication properties would be nice too (i.e set in 1 read from 5), as would a scale up/down example (kubectl edit the petset and modify replicas to 6, does the new replica catch up to the cluster?). Is there a "master" we should avoid killing during upgrade? 

Optionally, you could write something, bounce the whole cluster (delete pods --all), read the thing. 
Also mentioning read/write throughput and how that scales with replicas will help.

tschottdorf replied:
Done. I didn't go all out because `bash` isn't the right language for these tests (and we already have them elsewhere).
Also not too much about tradeoffs etc - I think for the toy example it's fine to assume that people generally have an idea of what they're toying with.

bprashanth commented on examples/cockroachdb/cockroachdb-petset.yaml:
```diff
@@ -0,0 +1,88 @@
+    app: cockroachdb
+spec:
+  ports:
+  - port: 26257
```
please add a comment elaborating what these ports are used for (pg protocol and raft, right?)

tschottdorf replied:
Not quite. I added a comment.

bprashanth commented on examples/cockroachdb/cockroachdb-petset.yaml:
```diff
@@ -0,0 +1,88 @@
+            # and smack it into our existing cluster.
+            # There are likely ways out. For example, the init container could
+            # query the kubernetes API and see whether any other nodes are
+            # around, etc. Or, of course, the admin can pre-seed the lost
```
please include a "limitations" section in the README above

tschottdorf replied:
Done.

smarterclayton replied:
This problem (it can also occur before a cluster has been initialized the first time) is something we need to solve by providing a strongly consistent config change mechanism available to nodes.  I think we'll make that a criteria for beta.

bprashanth commented on examples/cockroachdb/cockroachdb-petset.yaml:
```diff
@@ -0,0 +1,88 @@
+            fi
+            /cockroach/cockroach ${CRARGS[*]}
+      volumes:
+      - name: datadir
```
there are 2 ways to run this:
1. Assume people have dynamic provisioners
2. Make something that works out of the box

2 seems like the best approach for now, since dynamic provisioning is in alpha, but I suspect most people serious about this will run it with 1. Please include some text in the README saying something like "if you have a dynamic provisioner running in your cluster [link] you don't need to create the volumes".

tschottdorf replied:
2 puts me in unfamiliar terrain. Is there another example that demonstrates what you mean?

smarterclayton replied:
@thockin if we allowed EmptyDir to be added to PV then users could demonstrate this without gyrations today (the whole discussion about it being too hard to actually use PV for dev / test).

The alternative for now (which you could do here) is to have users create N hostPath PVs with a unique path per, and then hope that folks are running as root / have SELinux turned off.  We really need to document that.

bprashanth replied:
Or always deploy a default throwaway directory storage provisioner https://github.com/kubernetes/kubernetes/issues/27732

smarterclayton replied:
Yeah, but I'd rather do it with emptyDir.  I don't think tmpfs solves the
problem for medium sized directory sets.

Ultimately this is a storage class ("free") and a durability guarantee
("instance") and a quota (based on request capacity).  Since we already
have an alpha empty dir quota solution I want to sketch out the path that
gets us to where users *always* just create PVCs and they never have design
without them.

On Jul 4, 2016, at 9:06 PM, Prashanth B <notifications@github.com> wrote:

In examples/cockroachdb/cockroachdb-petset.yaml
<https://github.com/kubernetes/kubernetes/pull/28446#discussion_r69497879>:

> +            # Of course this isn't without danger - if node0 loses its data,
> +            # it will upon restarting it will simply bootstrap a new cluster
> +            # and smack it into our existing cluster.
> +            # There are likely ways out. For example, the init container could
> +            # query the kubernetes API and see whether any other nodes are
> +            # around, etc. Or, of course, the admin can pre-seed the lost
> +            # volume somehow (and in that case we should provide a better way,
> +            # for example a marker file).
> +            if [ ! "$(hostname)" == "cockroachdb-0" ] || \
> +               [ -e "/cockroach/cockroach-data/COCKROACHDB_VERSION" ]
> +            then
> +              CRARGS+=("--join" "cockroachdb")
> +            fi
> +            /cockroach/cockroach ${CRARGS[*]}
> +      volumes:
> +      - name: datadir

Or always deploy a default throwaway directory storage provisioner #27732
<https://github.com/kubernetes/kubernetes/issues/27732>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/28446/files/a402b08521bc41a113668d2f4f1fab5f10622302#r69497879>,
or mute the thread
<https://github.com/notifications/unsubscribe/ABG_p_fTiXcbkOHDG-WXI00TpfBCt2wnks5qSa4bgaJpZM4JEBSZ>
.

tschottdorf replied:
Anything left to do here? I'm waiting for feedback (read that PRs are
closed after 2wks of inactivity).

On Tue, Jul 5, 2016, 03:27 Clayton Coleman <notifications@github.com> wrote:

> In examples/cockroachdb/cockroachdb-petset.yaml
> <https://github.com/kubernetes/kubernetes/pull/28446#discussion_r69498643>
> :
>
> > +            # Of course this isn't without danger - if node0 loses its data,
> > +            # it will upon restarting it will simply bootstrap a new cluster
> > +            # and smack it into our existing cluster.
> > +            # There are likely ways out. For example, the init container could
> > +            # query the kubernetes API and see whether any other nodes are
> > +            # around, etc. Or, of course, the admin can pre-seed the lost
> > +            # volume somehow (and in that case we should provide a better way,
> > +            # for example a marker file).
> > +            if [ ! "$(hostname)" == "cockroachdb-0" ] || \
> > +               [ -e "/cockroach/cockroach-data/COCKROACHDB_VERSION" ]
> > +            then
> > +              CRARGS+=("--join" "cockroachdb")
> > +            fi
> > +            /cockroach/cockroach ${CRARGS[*]}
> > +      volumes:
> > +      - name: datadir
>
> Yeah, but I'd rather do it with emptyDir. I don't think tmpfs solves the
> problem for medium sized directory sets. Ultimately this is a storage class
> ("free") and a durability guarantee ("instance") and a quota (based on
> request capacity). Since we already have an alpha empty dir quota solution
> I want to sketch out the path that gets us to where users *always* just
> create PVCs and they never have design without them. On Jul 4, 2016, at
> 9:06 PM, Prashanth B <notifications@github.com> wrote: In
> examples/cockroachdb/cockroachdb-petset.yaml <#28446 (comment)
> <https://github.com/kubernetes/kubernetes/pull/28446#discussion_r69497879>
> >:
> + # Of course this isn't without danger - if node0 loses its data, + # it
> will upon restarting it will simply bootstrap a new cluster + # and smack
> it into our existing cluster. + # There are likely ways out. For example,
> the init container could + # query the kubernetes API and see whether any
> other nodes are + # around, etc. Or, of course, the admin can pre-seed the
> lost + # volume somehow (and in that case we should provide a better way, +
> # for example a marker file). + if [ ! "$(hostname)" == "cockroachdb-0" ]
> || \ + [ -e "/cockroach/cockroach-data/COCKROACHDB_VERSION" ] + then +
> CRARGS+=("--join" "cockroachdb") + fi + /cockroach/cockroach ${CRARGS[*]} +
> volumes: + - name: datadir
> Or always deploy a default throwaway directory storage provisioner #27732
> <https://github.com/kubernetes/kubernetes/issues/27732> <#27732
> <https://github.com/kubernetes/kubernetes/issues/27732>> — You are
> receiving this because you were mentioned. Reply to this email directly,
> view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/28446/files/a402b08521bc41a113668d2f4f1fab5f10622302#r69497879>,
> or mute the thread <
> https://github.com/notifications/unsubscribe/ABG_p_fTiXcbkOHDG-WXI00TpfBCt2wnks5qSa4bgaJpZM4JEBSZ>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28446/files/a402b08521bc41a113668d2f4f1fab5f10622302#r69498643>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE135M6Jf7QmzgU6JciYxe9S1-yVyLhSks5qSbLpgaJpZM4JEBSZ>
> .
>

bprashanth commented:
@tschottdorf thanks! @kubernetes/examples @smarterclayton

tschottdorf commented:
Updated, PTAL.

bprashanth commented on examples/cockroachdb/cockroachdb-petset.yaml:
```diff
@@ -0,0 +1,96 @@
+            # We will need some version of `cockroach init` back for this to
+            # work. For now, just do the same in a shell snippet.
+            # Of course this isn't without danger - if node0 loses its data,
+            # it will upon restarting it will simply bootstrap a new cluster
```
you have "it will" twice.

bprashanth commented:
ok to test

bprashanth commented:
@tschottdorf thanks for the example. I've marked it lgtm with just one nit. If you get to it before it goes in, great, if not we can fix it with a follow up. 

sorry for the delay, got sidetracked.

tschottdorf commented:
No worries, addressed your comment, rebased, and squashed.

On Wed, Jul 27, 2016 at 6:05 PM Prashanth B <notifications@github.com>
wrote:

> @tschottdorf <https://github.com/tschottdorf> thanks for the example.
> I've marked it lgtm with just one nit. If you get to it before it goes in,
> great, if not we can fix it with a follow up.
>
> sorry for the delay, got sidetracked.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28446#issuecomment-235735931>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135A_M89sLlyshpa0CIODxFiFGEa-vks5qZ9YsgaJpZM4JEBSZ>
> .
>

bprashanth commented:
Please run: 
./hack/verify-boilerplate.sh
and do as directed, I think jenkins verification failed on it (copyright might need fixing)

tschottdorf commented:
Yep, thanks. Will fix.

tschottdorf commented:
Done.

a-robinson commented:
@bprashanth, I have a couple questions around what it would take to flesh this out a little more once @tschottdorf has time to fix the doc munger's complaints:

1. Is there a good way to use secrets with petsets yet? Meaning if I have a certificate pre-created for each pet (i.e. cert-0, cert-1, etc.) and want the cert to end up with the pet that has the same index, is there a way to do that other than the pet manually grabbing it from the API when it starts up?
1. If we want clients of cockroach to spread their load among the pets (which are symmetric from the clients' point of view), I assume we'd have to create a second service with a ClusterIP? My understanding is that the headless service just returns a list endpoints, so we'd be depending on clients to properly randomize their usage of them if we relied on just the headless service.

bprashanth commented:
1. Secrets aren't a part of identity, we discussed it initially but there are other ways to distribute secrets (i.e generate something like a csr with your own hostname in the init container and hang on some pki service to sign it), and we don't autoprovision secrets like volumes so it's not something petset currently handles. 
2. The governing service is headless meaning it doesn't have clusterIP. You can lookup the name, and it'll return ips, the ordering of those ips will vary with the client (last I checked nslookup randomized, g's net.LookupHost didn't). If you want loadbalancing you need a service or ingress in front.

tbg committed:
Add a CockroachDB PetSet example

The example starts a simple five-node cluster with otherwise
default setting (in particular, 3x replication).

tschottdorf commented:
Munged the docs and pushed.

bprashanth commented:
ok to test
